src/bucket/BucketIndex.cpp

```diff
@@ -0,0 +1,241 @@
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "bucket/BucketIndex.h"
+#include "bucket/Bucket.h"
+#include "bucket/LedgerCmp.h"
+#include "main/Config.h"
+#include "util/Fs.h"
+#include "util/LogSlowExecution.h"
+#include "util/Logging.h"
+#include "util/XDRStream.h"
+
+#include <Tracy.hpp>
+#include <fmt/chrono.h>
+#include <fmt/format.h>
+#include <ledger/LedgerHashUtils.h>
+
+namespace stellar
+{
+
+// Returns a poolshare trustline key with the given accountID and a PoolID
+// filled with the fill byte
+static LedgerKey
+getDummyPoolShareTrustlineKey(AccountID const& accountID, uint8_t fill)
+{
+    LedgerKey key(TRUSTLINE);
+    key.trustLine().accountID = accountID;
+    key.trustLine().asset.type(ASSET_TYPE_POOL_SHARE);
+    key.trustLine().asset.liquidityPoolID().fill(fill);
+    return key;
+}
+
+template <class IndexT>
+BucketIndexImpl<IndexT>::BucketIndexImpl(std::filesystem::path const& filename,
+                                         std::streamoff pageSize)
+    : mPageSize(pageSize)
+{
+    ZoneScoped;
+
+    // Assert pageSize is a power of 2 for disk IO optimization
+    releaseAssert((pageSize & (pageSize - 1)) == 0);
+    releaseAssert(!filename.empty());
+
+    auto timer = LogSlowExecution("Indexing bucket");
+
+    size_t const estimatedLedgerEntrySize =
+        xdr::xdr_traits<BucketEntry>::serial_size(BucketEntry{});
+    auto fileSize = fs::size(filename);
+    auto estimatedNumElems = fileSize / estimatedLedgerEntrySize;
```

Comment:
It may not matter depending on how good the estimation is, but strictly more is better. According to the bloom filter library the estimate only needs to be within an order of magnitude of the actual number. However playing around with [this calculator](https://hur.st/bloomfilter), it looks like an underestimation of 1% increases the false positive rate by 7%, while an underestimation of 5% increases false positives by 40%. Overestimation results in more memory usage, but that is much more preferable to an increase in false positive rate. I don't think we should log bloom filter setup every time, but maybe just a warning if the error is larger than some percentage.

Reply:
I'll add this message as a WARNING for now if we underestimate. Maybe is should be INFO instead? I will definitely revisit this issue before BucketListDB is on by default, but for this experimental PR I'll just add the log message.